### PR TITLE
Refactor `Ansatz` to be a concrete type with lattice information

### PR DESCRIPTION
**Pull request by mofeing**
As discussed offline, we decided to slice #204 into smaller PRs. This first PR makes `Ansatz` a concrete type with lattice information contained in a graph.

The important thing you should review is the API, not the details; we can change the graph type in the future but we want the API to be stable.

# Changelog
- Introduced `Lattice` graph type
  - It's basically a `Graphs.SimpleGraph` with a bijective mapping between vertices and `Site`s
  - It adds `Graphs` ~and `MetaGraphsNext`~ as dependenc~ies~y
- Make `Ansatz` a concrete type in the type hierarchy
  - Its abstract counterpart is `AbstractAnsatz`
- Delete `Chain`, `Product`, `Grid` and `Dense`
  - These will be re-added in subsequent PRs, adapted to the new `Ansatz` layer
- Make `Form` the canonical form information trait
- Refactor some `evolve!` methods based on simple-update to `simple_update!` and generalize it for any `AbstractAnsatz`
- Added `@testset_skip` macro to implement @jofrevalles's idea of skipping whole testsets temporarily

# Notes
- ~Currently there is a bug in Julia makes Enzyme crash on an `Ansatz` due to its field `lattice` being abstract https://github.com/EnzymeAD/Enzyme.jl/issues/1923~ Should no longer be necessary because we replaced `MetaGraph` for `Lattice`
  - ~It's fixed in Julia 1.11.0 and should be fixed in Julia ~1.10.6~ 1.10.7~
  - ~I temporarily fixed it by fixing `Ansatz` to just 1D connectivity (i.e. `Site{1}`) but I'm removing it now -> For the time being you will need Julia 1.11 if you want to diff an `Ansatz`~
- ~There are other bug fixes (like `Base.in` or `GraphMakie.graphplot!` on `AbstractTensorNetwork`) that I'm gonna remove now and move to other PRs~ Already pushed to master
- Might change `Lattice` to `Topology` in the future, because it's more accurate if we're not using the partial ordering property of lattices
- In the future, we might change the subjacent graph type in `Lattice` for custom ones to accommodate clustered sites (i.e. a tensor containing 2+ sites) or hypergraphs

### CODE REVIEW

File: src/Tenet.jl [hunk 1]
```diff
@@ -26,24 +26,9 @@ include("Ansatz/Ansatz.jl")
 export Ansatz
 export socket, Scalar, State, Operator
 export boundary, Open, Periodic
+export form
```
Comment (jofrevalles) [inline src/Tenet.jl:30]:
```suggestion
```

**Review by jofrevalles — CHANGES_REQUESTED** [review]

File: src/Tenet.jl (hunk 1 above)
Comment (mofeing) [reply src/Tenet.jl:30]: why not export it?

**Comment by jofrevalles** [comment]
@mofeing @starsfordummies I think that when we talked about splitting the large PR into small ones, we meant that this PR would only contain the new `Ansatz` struct, and I see here that you pushed the commits regarding the canonical form, which IMO they should be in another PR. In this way, we can refactor the tests that are failing (like `Chain`, `Dense`, ...) in another PR.

**Comment by mofeing** [comment]
> @mofeing @starsfordummies I think that when we talked about splitting the large PR into small ones, we meant that this PR would only contain the new `Ansatz` struct, and I see here that you pushed the commits regarding the canonical form, which IMO they should be in another PR. In this way, we can refactor the tests that are failing (like `Chain`, `Dense`, ...) in another PR.

Sorry, but I don't agree. The `Form` trait is part of `Ansatz`, even if it doesn't have a field for that.

The other ansatzes will be reimplemented in subsequent PRs.

File: test/integration/YaoBlocks_test.jl [hunk 2]
```diff
@@ -8,35 +8,35 @@
     @test issetequal(sites(tn), [site"1", site"2", site"1'", site"2'"])
     @test Tenet.ntensors(tn) == 2
 
-    @testset "GHZ Circuit" begin
-        circuit_GHZ = chain(n_qubits, put(1 => Yao.H), Yao.control(1, 2 => Yao.X), Yao.control(2, 3 => Yao.X))
+    # @testset "GHZ Circuit" begin
```
Comment (jofrevalles) [inline test/integration/YaoBlocks_test.jl:11]:
```suggestion
    @test_skip begin #@testset "GHZ Circuit"
```

File: test/Product_test.jl [hunk 3]
```diff
@@ -1,33 +1,33 @@
-@testset "Product ansatz" begin
-    using LinearAlgebra
+# @testset "Product ansatz" begin
```
Comment (jofrevalles) [inline test/Product_test.jl:1]:
```suggestion
@test_skip begin # @testset "Product ansatz"
```

File: test/Ansatz_test.jl [hunk 4]
```diff
@@ -0,0 +1,2 @@
+@testset "Ansatz" begin
```
Comment (jofrevalles) [inline test/Ansatz_test.jl:1]: Add tests.

File: src/Ansatz/Ansatz.jl [hunk 5]
```diff
@@ -36,43 +81,280 @@ Base.show(io::IO, tn::A) where {A<:Ansatz} = summary(io, tn)
...
+
+Truncate the dimension of the virtual `bond`` of an [`Ansatz`](@ref) Tensor Network by keeping only the `maxdim` largest Schmidt coefficients or those larger than`threshold`.
 
-    return tn[only(inds(tensor1) ∩ inds(tensor2))]
+# Notes
+
+  - Either `threshold` or `maxdim` must be provided. If both are provided, `maxdim` is used.
+  - The bond must contain the Schmidt coefficients, i.e. a site canonization must be performed before calling `truncate!`.
+"""
+function truncate!(tn::AbstractAnsatz, bond; threshold=nothing, maxdim=nothing)
```
Comment (jofrevalles) [inline src/Ansatz/Ansatz.jl:123]: IMO we should have truncate only for `Canonical` or `MixedCanonical` now

**Review by jofrevalles — CHANGES_REQUESTED** [review]
I prefer to have `@test_skip` instead of commenting all the tests.

**Comment by mofeing** [comment]
> I prefer to have `@test_skip` instead of commenting all the tests.

That's a valid request and a good idea. Will change it now and recover the deleted tests as `@test_skip`... The only thing is I'm not sure if it will still try to evaluate the code, which will obviously crash because we're breaking the API. So maybe `@test_broken` but doesn't matter which.

File: test/Ansatz_test.jl (hunk 4 above)
Comment (mofeing) [reply test/Ansatz_test.jl:1]: I'm on it.

File: src/Ansatz/Ansatz.jl (hunk 5 above)
Comment (mofeing) [reply src/Ansatz/Ansatz.jl:123]: I don't agree. Canonizing a QTN can be costly and truncating a non-canonical TN is correctly defined. It is true that local truncation is far from optimal, but there are cases where you don't have an alternative or it's good enough (e.g. the graphene paper used local truncation without canonization and worked good enough).

File: test/integration/YaoBlocks_test.jl (hunk 2 above)
Comment (mofeing) [reply test/integration/YaoBlocks_test.jl:11]: I added a new `@testset_skip` macro for this.

File: test/runtests.jl [hunk 6]
```diff
@@ -10,8 +12,10 @@ using OMEinsum
     include("Transformations_test.jl")
     include("Site_test.jl")
     include("Quantum_test.jl")
-    include("Product_test.jl")
-    include("Chain_test.jl")
+    include("Lattice_test.jl")
+    include("Ansatz_test.jl")
+    # include("Product_test.jl")
+    # include("Chain_test.jl")
```
Comment (jofrevalles) [inline test/runtests.jl:18]:
```suggestion
    include("Product_test.jl")
    include("Chain_test.jl")
```
If we have `testskip` inside we don't need to comment out this, right?

File: src/Ansatz.jl [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+Truncate the dimension of the virtual `bond`` of an [`Ansatz`](@ref) Tensor Network by keeping only the `maxdim` largest Schmidt coefficients or those larger than`threshold`.
+
+# Notes
+
+  - Either `threshold` or `maxdim` must be provided. If both are provided, `maxdim` is used.
+  - The bond must contain the Schmidt coefficients, i.e. a site canonization must be performed before calling `truncate!`.
+"""
+function truncate!(tn::AbstractAnsatz, bond; threshold=nothing, maxdim=nothing)
+    @assert isnothing(maxdim) ⊻ isnothing(threshold) "Either `threshold` or `maxdim` must be provided"
+
```
Comment (jofrevalles) [inline src/Ansatz.jl:217]:
```suggestion
    @assert form(ψ) != NonCanonical() "A `NonCanonical` tensor network can not be truncated"

```

**Review by jofrevalles — CHANGES_REQUESTED** [review]

File: src/Ansatz.jl (hunk 7 above)
Comment (mofeing) [reply src/Ansatz.jl:217]: Again, I don't agree with this affirmation. You CAN truncate even if it's far from optimal, and in many cases it works.

**Comment by jofrevalles** [comment]
If we fix the documentation I guess we can merge

**Comment by mofeing** [comment]
> If we fix the documentation I guess we can merge

Docs are basically broken. We need to redo them but after this PR.

**Review by jofrevalles — APPROVED** [review]

**Review by starsfordummies — APPROVED** [review]
Sergio breaks tenet in all possible ways without even documenting what's broken. Typical commit, looks good to me 👍

**Review by arturgs — APPROVED** [review]

**Pull request merged**